[Mono] Fix all the issues related to enabling WasmStripILAfterAOT

fanyang-mono opened the pull request:
This PR enables IL trim by default for WASM apps, by solving the following major problems:

- There are various opcode which was translated into a function call. I modified the `call` goto logic to check if there is an AOT'ed version of the callee method. If there is, invoke that. If there isn't, interp compile the callee method.
- Move the logic of checking if a method is supported a JIT/AOT call from interpreter callback to `mini-runtime.c`. Because the callback was initialized with the desired logic only when `mono_use_interpreter` is set to true. See the following code 
https://github.com/dotnet/runtime/blob/8d4f8b86a742caba0f7640c013a2c1cefe10b146/src/mono/mono/mini/mini-runtime.c#L4629-L4632
However, in `driver.c`, `mono_use_interpreter` was not set to true during AOT compile time, even though `interp` was passed in as an AOT option. It would be set to true during AOT runtime. Alter this behavior is probably not desirable. Thus, I choose to move the checking logic to `mini-runtime.c`
- Set the trimmed assemblies with the same timestamp as the original assemblies. So that it won't trigger unwanted rebuild or re-AOT
- Added implementation of `g_fopen` to properly handle unicode string on windows. 
- Interpreter can't do JIT call for methods which require synchronization. Remove those methods from trimmable method list.
- Disabled tailcall optimization when calling trimmed methods.
- Do jitcall when encountering virtual tailcalls.
- Change of ILStrip task:
  - The trimmed assemblies were saved in a new folder with the same name as the original assembly.
  - Created an output variable `UpdatedAssemblies` with the updated path of assemblies. The trimmed assemblies are updated with their new path.
  - Updated the WASM build toolchain to replace the content of `_WasmAssembliesInternal` with `UpdatedAssemblies`
- Added the input parameter `WasmRuntimeConfigFilePath` for `WasmAppBuilderBaseTask`. Got rid of the logic to calculate the path of `runtimeconfig.json` file. Use the value of `WasmRuntimeConfigFilePath` instead. (Fixes #93692)
- Found two issues related to wasm build process. @radical fixed them.
  - #93680
  - Fixes #93689
- Blazor app is able to packages the trimmed assemblies, when `WasmStripILAfterAOT` is enabled.
- Added tests for wasm template app and blazor template app 

**Library size change for top 3 largest libraries:**
- System.Private.CoreLib.dll.br: 452k -> 316K
- System.Text.Json.dll.br: 109k -> 81k
- Microsoft.AspNetCore.Components.dll.br: 69k -> 54k

fanyang-mono committed:
Enable IL trim for WASM by default

SamMonoRT commented:
cc @lewing @kg

kg commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -149,7 +149,7 @@
     <WasmAssemblyExtension Condition="'$(WasmEnableWebcil)' != 'true'">.dll</WasmAssemblyExtension>
 
     <!-- by default, method body IL code trimming is disabled -->
```
make sure to update the comment

fanyang-mono replied:
updated :)

SamMonoRT commented:
As discussed, merge this only after RC1 has forked and discussions with @lewing

fanyang-mono committed:
Make ILStrip available for local build

fanyang-mono committed:
Make all calling another method go through the logic to see if it could call an AOT'ed version of it before trying to interp compile it

fanyang-mono commented on src/mono/mono/mini/aot-compiler.c:
```diff
@@ -9868,7 +9868,8 @@ compile_method (MonoAotCompile *acfg, MonoMethod *method)
 			// The call back to jit_call_can_be_supported is necessary for WASM, because it would still interprete some methods sometimes even though they were already AOT'ed.
 			// When that happens, interpreter needs to have the capability to call the AOT'ed version of that method, since the method body has already been trimmed.
-			fprintf (acfg->trimming_eligible_methods_outfile, "%x\n", method->token);
+			if ((strstr(mono_method_full_name(method, TRUE), "System.Type:GetType") == NULL) && (strstr(mono_method_full_name(method, TRUE), "System.Tests.DelegateTests:AllPrimitivesMethod") == NULL))
```
This is a hack. Will work on removing it before open for review.

fanyang-mono committed:
Add back accidentally removed line of code

fanyang-mono committed:
Update test to accommodate IL trim with WASM AOT

fanyang-mono committed:
Move jit_call_can_be_supported to mini-runtime, so it doesn't depends on the value of mono_use_interpreter

fanyang-mono committed:
Update var name

fanyang-mono committed:
Attempt to fix rebuild test failures

vargaz commented on src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs:
```diff
@@ -308,5 +308,9 @@ private void AddItemToTrimmedList(string assemblyFilePath, string trimmedAssembl
         trimmedAssemblyItem.SetMetadata("TrimmedAssemblyFileName", trimmedAssemblyFilePath);
         _trimmedAssemblies.Add(trimmedAssemblyItem);
+
+        FileInfo fileInfo = new FileInfo(assemblyFilePath);
```
Why is this needed ?

fanyang-mono replied:
The wasm rebuild tests failed, when IL trim is enabled. Test failure log can be found at https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-239a4747ba51427c95/Workloads-Wasm.Build.NativeRebuild.Tests.FlagsChangeRebuildTests/1/console.3814634a.log?helixlogtype=result

It seems that the rebuild mechanism detected that the assembly files are newer than the AOT results. So it tried to re-do AOT. This will hopefully stop that.

vargaz replied:
This sets the timestamp of the trimmed assembly file which is unrelated to AOT.

fanyang-mono replied:
The trimmed assembly files later will replace the original assembly files, which were AOT compiled before being trimmed.

radical replied:
I hope (🤞) this won't break anything else. Could you please add a comment here explaining the reasoning?

fanyang-mono replied:
Added comment :)

fanyang-mono committed:
Attempt to fix the file open issue with unicode on windows

fanyang-mono committed:
Attempt to fix unicode issue #2

radical commented:
/azp run runtime-wasm

fanyang-mono committed:
Enable g_fopen to have the capability of handling opening files with unicode name on all platforms

fanyang-mono committed:
Add comment

fanyang-mono committed:
Update comment

fanyang-mono committed:
Fix file indentation format

fanyang-mono committed:
Check if string contains non-ascii char

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono commented:
@lateralusX I would like you to review my implementation of `g_fopen`.

fanyang-mono committed:
Remove unused callback

vargaz commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -148,8 +148,8 @@
-    <!-- by default, method body IL code trimming is disabled -->
-    <WasmStripILAfterAOT Condition="'$(WasmStripILAfterAOT)' == ''">false</WasmStripILAfterAOT>
+    <!-- by default, method body IL code trimming is enabled -->
+    <WasmStripILAfterAOT Condition="'$(WasmStripILAfterAOT)' == ''">true</WasmStripILAfterAOT>
```
This comment is redundant.

fanyang-mono replied:
removed

fanyang-mono committed:
Remove redundant comment

vargaz commented on src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs:
```diff
@@ -308,5 +308,11 @@ private void AddItemToTrimmedList(string assemblyFilePath, string trimmedAssembl
+        FileInfo fileInfo = new FileInfo(assemblyFilePath);
+        DateTime lastModified = fileInfo.LastWriteTime;
+        File.SetLastWriteTime(trimmedAssemblyFilePath, lastModified);
     }
```
Still don't think this is the right way, in a build system, build outputs like these should have later timestamps than the build inputs.

fanyang-mono replied:
Without this, AOT generated bitcode files would have older timestamp than the assembly files. Then rebuild will trigger re-AOT, cause it thinks those assembly files have changed.

vargaz replied:
That seems to be a problem with other parts of the build which overwrites build files, its not possible to have correct build dependencies when files are overwritten.

fanyang-mono replied:
TBH, I am not sure why other build files were overwritten but not triggering rebuild. This did fix the test failures tho. Do you have suggestions on fixing it in some other way?

radical replied:
Instead of overwriting the original files the item `_WasmAssembliesInternal` can be modified to have the updated paths.

fanyang-mono replied:
Right after the IL strip task, the original assemblies will be replaces with the trimmed ones. The file paths didn't change.

radical replied:
I'm suggesting to not replace the original assembly files, and instead just change the paths to new ones in `_WasmAssembliesInternal`. That way rest of the build will be using the newly-trimmed assemblies, instead of the original ones.

radical replied:
but if we don't replace the original ones then on a rebuild it should not trigger a rebuild. And this should ensure that we have the correct build dependencies, like Zoltan was talking about.

fanyang-mono replied:
Oh I see what you mean. My worry for this is how we could guarantee tasks run after ILStrip step would always refer to the new ones?

radical replied:
`_WasmAssembliesInternal` is used by all the targets, that's the convention used in `WasmApp.Native.targets`, so that would ensure that the new ones are used.

radical replied:
We should fix this before merging.

fanyang-mono replied:
Fixed.

vargaz commented on src/mono/mono/mini/mini-runtime.c:
```diff
@@ -5555,3 +5555,28 @@ mono_invoke_runtime_init_callback (void)
+
+gboolean
+jit_call_can_be_supported_by_interp (MonoMethod *method, MonoMethodSignature *sig, gboolean is_llvm_only)
+{
```
This should have a mono_ or mini_ prefix.

fanyang-mono replied:
Modified.

fanyang-mono committed:
Update method name

fanyang-mono commented:
/azp run runtime-wasm

lateralusX commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+		return NULL;
+
+#ifndef HOST_WIN32
+	fp = fopen(path, mode);
```
```suggestion
	fp = fopen (path, mode);
```

fanyang-mono replied:
fixed

lateralusX commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+	fp = fopen(path, mode);
+#else
+	gboolean all_ascii = TRUE;
+	for (int i = 0; path[i] != '\0'; i++) {
```
```suggestion
	for (int i = 0; path [i] != '\0'; i++) {
```

lateralusX commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+		fp = fopen(path, mode);
+	} else {
+		gunichar2 *wPath = g_utf8_to_utf16(path, -1, 0, 0, 0);
+		gunichar2 *wMode = g_utf8_to_utf16(mode, -1, 0, 0, 0);
```
```suggestion
		gunichar2 *wMode = g_utf8_to_utf16 (mode, -1, 0, 0, 0);
```

fanyang-mono replied:
fixed

lateralusX reviewed (commented):
Just looked at the g_fopen, and it looks OK except a couple of formatting nits.

fanyang-mono commented:
CI test failures are relevant.

radical commented:
`MonoAOTCompiler` task should emit a separate item with the list of the token files, and their corresponding assemblies. This will make it clear in the .targets file what *input* is being ingested by `ILStrip`. Otherwise, it's magical.

radical commented:
Just curious - does the aot compiler remove (or zero-out) methods too? I mean, other than what the illinker does?

fanyang-mono commented:
> Just curious - does the aot compiler remove (or zero-out) methods too? I mean, other than what the illinker does?

Not to my knowledge.

radical commented:
> > `MonoAOTCompiler` task should emit a separate item with the list of the token files, and their corresponding assemblies. This will make it clear in the .targets file what _input_ is being ingested by `ILStrip`. Otherwise, it's magical.
> 
> It does. It is one of the metadata of the output `CompiledAssemblies`.

Yeah, but that's not clear from reading the targets. You need to look at the source of the task to find that out. Having inputs/outputs clear would improve readability too.

radical commented:
> > Just curious - does the aot compiler remove (or zero-out) methods too? I mean, other than what the illinker does?
> 
> Not to my knowledge.

In that case can we run the ILStrip step before `MonoAOTCompiler`?

radical commented:
> > > Just curious - does the aot compiler remove (or zero-out) methods too? I mean, other than what the illinker does?
> > 
> > 
> > Not to my knowledge.
> 
> In that case can we run the ILStrip step before `MonoAOTCompiler`?

Actually, we should be running this for relinking case also.

fanyang-mono commented:
> > > > Just curious - does the aot compiler remove (or zero-out) methods too? I mean, other than what the illinker does?
> > > 
> > > 
> > > Not to my knowledge.
> > 
> > 
> > In that case can we run the ILStrip step before `MonoAOTCompiler`?
> 
> Actually, we should be running this for relinking case also.

No, ILStrip step can't be run before `MonoAOTCompiler`. It needs to know which methods were AOT'ed, in order to trim it.
What is relinking case?

fanyang-mono committed:
Fixed some method not found issues and remove the optimization for g_fopen

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+		if (!wPath || !wMode)
+			return NULL;
+
+		fp = _wfopen((wchar_t *) wPath, (wchar_t *) wMode);
```
fixed

fanyang-mono commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+	if (all_ascii) {
+		fp = fopen(path, mode);
+	} else {
+		gunichar2 *wPath = g_utf8_to_utf16(path, -1, 0, 0, 0);
```
fixed

fanyang-mono commented on src/mono/mono/eglib/gfile.c:
```diff
@@ -155,3 +155,40 @@ g_file_set_contents (const gchar *filename, const gchar *contents, gssize length
+	}
+
+	if (all_ascii) {
+		fp = fopen(path, mode);
```
fixed

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -4259,7 +4195,45 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
 			ip += 4;
 #endif
-call:
+
```
The common call path (for `MINT_CALL`) shouldn't execute this code. When we do a static call we know whether the method to be called is aot or not so we are just adding unnecessary checks to a hot path.

fanyang-mono replied:
Moved it away from `MINT_CALL`.

fanyang-mono committed:
Fix tailcall

vargaz commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -3963,6 +3963,39 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
 				memmove (frame->stack, (guchar*)frame->stack + params_offset, params_size);
 				new_method = (InterpMethod*)frame->imethod->data_items [ip [2]];
 
+				{
```
This should be done in transform.c i think, the only time this needs to happen in interp.c is when doing indirect calls.

fanyang-mono replied:
Fixed.

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Disable tailcall optimization when calling a trimmed method

fanyang-mono commented:
/azp run runtime-wasm

vargaz commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3549,10 +3549,16 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
+	MonoMethodHeader *mheader = NULL;
+	if (tailcall && target_method != NULL)
+		mheader = interp_method_get_header (target_method, error);
+
```
This needs to be freed.

fanyang-mono replied:
Freed.

fanyang-mono committed:
Free method header

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Fix windows build error

fanyang-mono committed:
Free method header at the correct locatioin

fanyang-mono committed:
Fix the condition of skipping tailcall

fanyang-mono committed:
Fix test failure

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Move JIT/AOT call invoke away from MINT_CALL, as it is not needed there.

fanyang-mono commented:
/azp run runtime-wasm

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -4205,39 +4173,41 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
 			}
 
-			InterpMethodCodeType code_type = cmethod->code_type;
+jit_call:		
```
I would keep an eye open for potential regressions on wasm due to more complicated control flow which might hinder some compiler optimizations.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -3549,10 +3549,20 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
 	CHECK_STACK_RET (td, csignature->param_count + csignature->hasthis, FALSE);
+
+	gboolean skip_tailcall = FALSE;
+	if (tailcall && target_method != NULL) {
```
The virtual case is not really correctly handled by this. If we change the condition to `if (tailcall && !is_virtual && target_method != NULL) {`, would it still fix the crashes, for now ?

fanyang-mono replied:
Fixed :)

fanyang-mono committed:
Fix virtual tail call

fanyang-mono committed:
Merge remote-tracking branch 'origin/main' into enable_il_trim_wasm_always

fanyang-mono committed:
Merge branch 'main' into enable_il_trim_wasm_always

fanyang-mono commented:
/azp run runtime-wasm

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -3945,6 +3945,36 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
+					if (code_type == IMETHOD_CODE_COMPILED) {
+						frame->imethod = new_method;
+						ip = frame->imethod->code;
+						frame->state.ip = ip;
```
The code here is somewhat dubious to me. Since we determined that the method that we wanted to tailcall is not interp method, then we can no longer do a tailcall. So we shouldn't overwrite the contents of the calling frame. It should be like we just do a jitcall and then exit the frame. The memmove above shouldn't happen if we do a jitcall and we shouldn't really update `frame->imethod`, and `frame->state.ip`. It probably doesn't make a difference in practice, but the code might make more sense.

fanyang-mono replied:
Fixed.

fanyang-mono committed:
Address review feedback

fanyang-mono committed:
Merge branch 'main' into enable_il_trim_wasm_always

radical commented:
> [helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-eacc2608043a4c9fbe/Workloads-NoWebcil-Wasm.Build.Tests.Blazor.NativeTests/1/console.04e448b1.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-eacc2608043a4c9fbe/Workloads-NoWebcil-Wasm.Build.Tests.Blazor.NativeTests/1/console.04e448b1.log?helixlogtype=result) It seems that the reason why the test (Wasm.Build.Tests.Blazor.NativeTests.BlazorWasm_CanRunMonoAOTCross_WithNoTrimming(config: "Debug")) failed is because, the build artifacts from previous test (Wasm.Build.Tests.Blazor.NativeTests.BlazorWasm_CanRunMonoAOTCross_WithNoTrimming(config: "Release")) wasn't removed. It was trying to AOT the trimmed libraries.
> 
> @radical this seems to be a test issue?

The two tests are using different project paths - `blazorwasm_Release_aot_apnd5cks_tbs`, and `blazorwasm_Debug_aot_43u5mwdb_b4z`.

Also, one of the things to note for this test is that it does *not* use trimming, thus the paths that you get for AOT are from the runtime pack, and nugets. But the `Move` after `ILStrip` goes and overwrites those files, thus affecting any tests that follow!

fanyang-mono committed:
Put the trimmed assemblies in a new folder, output an updated list of assemblies and update _WasmAssembliesInternal with the new list

fanyang-mono commented:
> > [helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-eacc2608043a4c9fbe/Workloads-NoWebcil-Wasm.Build.Tests.Blazor.NativeTests/1/console.04e448b1.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-eacc2608043a4c9fbe/Workloads-NoWebcil-Wasm.Build.Tests.Blazor.NativeTests/1/console.04e448b1.log?helixlogtype=result) It seems that the reason why the test (Wasm.Build.Tests.Blazor.NativeTests.BlazorWasm_CanRunMonoAOTCross_WithNoTrimming(config: "Debug")) failed is because, the build artifacts from previous test (Wasm.Build.Tests.Blazor.NativeTests.BlazorWasm_CanRunMonoAOTCross_WithNoTrimming(config: "Release")) wasn't removed. It was trying to AOT the trimmed libraries.
> > @radical this seems to be a test issue?
> 
> The two tests are using different project paths - `blazorwasm_Release_aot_apnd5cks_tbs`, and `blazorwasm_Debug_aot_43u5mwdb_b4z`.
> 
> Also, one of the things to note for this test is that it does _not_ use trimming, thus the paths that you get for AOT are from the runtime pack, and nugets. But the `Move` after `ILStrip` overwrites those files, thus affecting any tests that follow!

Then my latest commit should be able to fix it.

fanyang-mono commented:
@kg @lewing @radical @vargaz @BrzVlad I believe that I've addressed all the feedback. Please do a final round of review.

This PR will enable `WasmStripILAfterAOT` by default. Please let me know if you have any concerns regarding this change of behavior. Thanks :)

fanyang-mono commented:
/azp run runtime-wasm

radical commented on src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs:
```diff
@@ -167,25 +184,32 @@ private bool TrimMethods(ITaskItem assemblyItem)
         else
         {
-            return Path.Combine(assemblyPath, (assemblyName + "_trimmed.dll"));
+            return Path.Combine(assemblyPath,  "trimmedAssemblies");
```
All of the trimmed assemblies should go into one directory, which is in our intermediate output directory.

The above would fail or be bad if the assembly path happened to be from a nuget package, or the runtime pack. Similar to the no-trimming test that was failing.

radical replied:
ILLinker uses a `linked/` for example.

fanyang-mono replied:
Are you suggesting `trimmed` instead of `trimmedAssemblies`?

radical replied:
I wasn't suggesting what the folder name should be - `trimmed` would be preferable, but either is fine really.

I was saying that with the current code the `trimmedAssemblies` directory is created next to the original assembly itself, which be a nuget cache path, or a runtime pack path etc. Instead of that, create one `trimmed` or `trimmedAssemblies` path in `obj` (WasmIntermediateOutputPath), and emit all the trimmed assemblies there. Easy to clean up too.

fanyang-mono replied:
Fixed.

radical commented on src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs:
```diff
@@ -167,25 +184,32 @@ private bool TrimMethods(ITaskItem assemblyItem)
+            newAssmeblyItem.ItemSpec = trimmedAssemblyFilePath;
         }
 
+        _updatedAssemblies.Add(newAssmeblyItem);
```
Instead of another item, you can make `UpdatedAssemblies` the only output item. And set a metadata on the item like `ILStripped=true`. This way it is a clean output. And if caller does need to differentiate somehow then they have a simple way of doing that.
Also, in case of a trimmed assembly add `UntrimmedAssemblyPath` metadata to the item. This will help in any debugging, or if the caller needs the original path too.

fanyang-mono replied:
The name of the file path should indicate it is a trimmed assembly.

radical replied:
The targets might need to differentiate. It is a general convention in msbuild to add such metadata which can help with filtering. Otherwise, any targets would need to assume the path structure.

fanyang-mono replied:
Updated.

vargaz commented:
The non-interpreter changes look ok to me.

fanyang-mono committed:
Put trimmed assemblies in IntermediateOutputPath

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Remove TrimmedAssemblies

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Create trimmed assembly folder before the parallel run

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Try to fix the issue with missing item

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Fix parallelism issue

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Merge branch 'main' into enable_il_trim_wasm_always

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono commented:
@radical wasm build test failed with this error
```
      CompareStat failed:
      [Expected unchanged file: dotnet.native.wasm]
         old: FileStat { Exists = True, LastWriteTimeUtc = 10/12/2023 10:20:33 PM, Length = 8211204, FullPath = /root/helix/work/workitem/e/wbt/xohakygu_lhm/bin/Release/net8.0/browser-wasm/AppBundle/_framework/dotnet.native.wasm }
         new: FileStat { Exists = True, LastWriteTimeUtc = 10/12/2023 10:21:16 PM, Length = 8199175, FullPath = /root/helix/work/workitem/e/wbt/xohakygu_lhm/bin/Release/net8.0/browser-wasm/AppBundle/_framework/dotnet.native.wasm }
      [Expected unchanged file: dotnet.native.js]
         old: FileStat { Exists = True, LastWriteTimeUtc = 10/12/2023 10:20:30 PM, Length = 247415, FullPath = /root/helix/work/workitem/e/wbt/xohakygu_lhm/bin/Release/net8.0/browser-wasm/AppBundle/_framework/dotnet.native.js }
         new: FileStat { Exists = True, LastWriteTimeUtc = 10/12/2023 10:21:13 PM, Length = 247415, FullPath = /root/helix/work/workitem/e/wbt/xohakygu_lhm/bin/Release/net8.0/browser-wasm/AppBundle/_framework/dotnet.native.js }
```
I am wondering if you know what might went wrong?

Here is the full log: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-90436-merge-69566945f848418fa5/Workloads-Wasm.Build.NativeRebuild.Tests.NoopNativeRebuildTest/1/console.e623f4c8.log?helixlogtype=result

vargaz commented:
Seems to mean that the build is not incremental enough, i.e. output files are changed even if the inputs files are the same.

radical commented:
> Seems to mean that the build is not incremental enough, i.e. output files are changed even if the inputs files are the same.

That's correct.

fanyang-mono commented:
@radical Do you know what might have caused it and a possible solution?

radical commented:
> @radical Do you know what might have caused it and a possible solution?

The `dotnet.native.{js,wasm}` files have changed. That would happen when relinking. I would suggest looking at the binlog for that project build, and check why it didn't get skipped.

```
       []   AOT'ing 3 assemblies
        []   Everything is up-to-date, nothing to precompile
        []   Linking for initial memory $(EmccInitialHeapSize)=16777216 bytes. Set this msbuild property to change the value.
        []   Linking with emcc with -O2. This may take a while ...
```

That `linking with .. ` shouldn't have happened.

fanyang-mono commented:
@radical Could you help me with investigating this issue? I am not familiar with Wasm build process, thus not sure what to look for in the binlog.

fanyang-mono committed:
Only start the trim when the assembly is newer than the output

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Add assembly item to the list, when

radical commented on src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs:
```diff
@@ -143,7 +171,21 @@ private bool TrimMethods(ITaskItem assemblyItem)
+        string trimmedAssemblyFilePath = ComputeTrimmedAssemblyPath(trimmedAssemblyFolder, assemblyFilePath);
+        if (File.Exists(trimmedAssemblyFilePath))
+        {
+            if (IsInputNewerThanOutput(assemblyFilePath, trimmedAssemblyFilePath))
```
Please log whether the file is being trimmed or being skipped, and why. See https://github.com/dotnet/runtime/blob/25961d1156b3ed8cf684aaa9c56d78a8b2cecdfe/src/tasks/WasmAppBuilder/EmccCompile.cs#L251-L296 for an example.

fanyang-mono replied:
Added.

fanyang-mono committed:
Add some logging

fanyang-mono committed:
Fixed runtimeconfig.json file path issue and disabed failed tests

fanyang-mono commented:
/azp run runtime-wasm

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilderBaseTask.cs:
```diff
@@ -22,6 +22,10 @@ public abstract class WasmAppBuilderBaseTask : Task
 
+    [NotNull]
+    [Required]
+    public string? WasmRuntimeConfigFilePath { get; set; }
```
rename to `RuntimeConfigJsonPath`

fanyang-mono replied:
Updated.

fanyang-mono committed:
Update parameter name

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Fix wasi build

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Use the correct parameter

fanyang-mono committed:
Merge branch 'main' into enable_il_trim_wasm_always

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Fix runtime test failure

fanyang-mono commented:
/azp run runtime-wasm

radical committed:
Merge remote-tracking branch 'origin/main' into enable_il_trim_wasm_always

radical committed:
WasmAppBuilder: runtimeconfig.json path can be null

radical committed:
cleanup

radical committed:
ILStrip: fix typo in id name

radical committed:
Cleanup

radical committed:
ILStrip: ensure output assemblies are in the same order as the input. This is required for incremental builds.

radical committed:
more cleanup

radical committed:
Re-enable disabled tests

radical commented:
/azp run runtime-wasm

radical reviewed (dismissed):
Build and wbt changes look good 👍

BrzVlad reviewed (approved):
interp looks good

fanyang-mono committed:
Change the default value for WasmStripILAfterAOT to false

fanyang-mono commented:
/azp run runtime-wasm

lewing commented:
I edited the title because Il trimming is not striping or zeroing

maraf commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -712,14 +712,15 @@
-    />
+
+    <ItemGroup>
+      <_WasmAssembliesInternal Remove="@(_WasmAssembliesInternal)" Condition="'$(WasmStripILAfterAOT)' == 'true'" />
```
nit: The condition can be moved to `ItemGroup`

fanyang-mono replied:
Updated.

maraf commented:
If I build an app with
```
dotnet publish -bl -p:RunAOTCompilation=true
```
and than without any change run
```
dotnet.exe publish -bl -p:RunAOTCompilation=true -p:WasmStripILAfterAOT=true
...
Everything is up-to-date, nothing to precompile
```
I'm getting errors like (tokens folder is empty)
```
....\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\9.0.0-dev\Sdk\WasmApp.Native.targets(712,5): error : ....\BlazorILStrip\obj\Release\net8.0\wasm\for-publish\tokens\Microsoft_Extensions_Configuration_Json_dll_
compiled_methods.txt doesn't exist. [....\BlazorILStrip\BlazorILStrip.csproj]
```

maraf commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -712,14 +712,15 @@
+
+    <ItemGroup>
+      <_WasmAssembliesInternal Remove="@(_WasmAssembliesInternal)" Condition="'$(WasmStripILAfterAOT)' == 'true'" />
+      <_WasmAssembliesInternal Include="@(_UpdatedAssembliesAfterILStrip)" Condition="'$(WasmStripILAfterAOT)' == 'true'" />
```
Blazor was never using `@(WasmAssembliesFinal)` and instead rely on `@(ResolvedFileToPublish)`. The AOT compiler works "in place" and so the `@(ResolvedFileToPublish)` contains correct items. We should update `@(ResolvedFileToPublish)` at the end of wasm publish

radical replied:
WasmSdk is selecting the files to be bundled here via the `@(WasmAssembliesToBundle)` item which is passed to `WasmApp.targets`:
https://github.com/dotnet/runtime/blob/f2e4ddf20936b37445bbbb2d139e8080ca00f16c/src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets#L595-L600

A corresponding target in Browser.targets should do the "unwrapping", and update whatever items needs to be updated.

maraf replied:
It didn't have to the unwrapping so just because this assembly location did not changed during AOT, right? 

The flow is that SWA (StaticWebAssets) has dependency on `ProcessPublishFilesForWasm` which triggers `WasmTriggerPublishApp` which triggers `_GatherWasmFilesToPublish` and that copies `@(ResolvedFileToPublish)` to `@(WasmAssembliesToBundle)`. Than neste publish happens and produces `@(WasmAssembilesFinal)`, which has the exact locations as original `@(ResolvedFileToPublish)`. Than finally the `ProcessPublishFilesForWasm` is exected and it uses `@(ResolvedFileToPublish)` to update SWA entries with AOTed assemblies. You suggest to replace the last by reading `@(WasmAssembliesFinal)` instead of `@(ResolvedFileToPublish)`, right?

I'm thinking, why do we need the copying to `WasmAssemblies**`? Why don't we use `ResolvedFileToPublish` for communication between `WasmApp.targets` and `Browser.targets`?

radical replied:
> I'm thinking, why do we need the copying to WasmAssemblies**? Why don't we use ResolvedFileToPublish for communication between WasmApp.targets and Browser.targets?

My original plan was to pick up publish artifacts automatically, and figure out "for just bundling", "assemblies", etc. But we also want to move away from `publish` as such, which would be in the sdk first. So, for now using the `WasmAssemblies*` allows us separation from the publish details. Also, `ResolvedFileToPublish` can have metadata to consider once we start adding that.

I also want to drop the nested publish now, since we have better control over what blazor does about the inputs.

fanyang-mono replied:
Do we have an agreement regarding to what should be the right thing to do here?

fanyang-mono replied:
@maraf if you two agreed on the solution. Do you mind pushing a change to this PR?

maraf replied:
We agreed to use `WasmAssembliesFinal` at the moment, before we drop nested publish. I'll try to update the PR today or at the latest on Monday

fanyang-mono committed:
Fix the issue with changing the value of WasmStripILAfterAOT between incremental builds

fanyang-mono committed:
Move the location of file deleting

fanyang-mono commented:
> If I build an app with
> 
> ```
> dotnet publish -bl -p:RunAOTCompilation=true
> ```
> 
> and than without any change run
> 
> ```
> dotnet.exe publish -bl -p:RunAOTCompilation=true -p:WasmStripILAfterAOT=true
> ...
> Everything is up-to-date, nothing to precompile
> ```
> 
> I'm getting errors like (tokens folder is empty)
> 
> ```
> ....\dotnet-latest\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\9.0.0-dev\Sdk\WasmApp.Native.targets(712,5): error : ....\BlazorILStrip\obj\Release\net8.0\wasm\for-publish\tokens\Microsoft_Extensions_Configuration_Json_dll_
> compiled_methods.txt doesn't exist. [....\BlazorILStrip\BlazorILStrip.csproj]
> ```

This should be fixed now.

maraf committed:
Use WasmAssembliesFinal or ResolvedFileToPublish during publish in WasmSDK

maraf commented:
We should have a Wasm Build Test for IL stripping and check that publish picks correct files

fanyang-mono commented:
> We should have a Wasm Build Test for IL stripping and check that publish picks correct files, but it can be done in a follow up

I am working on adding the tests.

fanyang-mono commented:
@maraf Could you take a look at the test failures. It's probably related to your change. One of the error messages is
```
Expected the file to exist: /root/helix/work/workitem/e/wbt/blz_resources_Debug_jghr3wqy_ei0/bin/Debug/net9.0/publish/wwwroot/_framework/System.Private.CoreLib.wasm
```

maraf commented:
Yes, it does. Sorry for that. I dunno why it didn't manifested when I tested locally. I'll fix it on Monday. Feel free to revert my commit it blocks you now.
Note to myself: WasmAssembliesFinal don't have RelativePath metadat set.

maraf committed:
Copy metadata in AOT compiler and when creating WasmAssembliesToBundle from ResolvedFileToPublish

maraf commented:
@radical `@(WasmAssembliesFinal)` don't contain satellite assemblies. They removed in https://github.com/dotnet/runtime/blob/6066de1e8e37e5befc3d04909fb522a37e5de940/src/mono/wasm/build/WasmApp.targets#L347

Do you prefer to add them back when `@(WasmAssembliesFinal)` are constructed from `@(_WasmAssembliesInternal)` or should I use `@(_WasmSatelliteAssemblies)` in WasmSDK targets?

maraf committed:
Add _WasmSatelliteAssemblies to WasmAssembliesFinal

fanyang-mono committed:
Add a wasm template test

maraf committed:
Include all non-dll ResolvedFileToPublish for ComputeWasmPublishAssets

maraf committed:
Merge remote-tracking branch 'upstream/main' into fan/enable_il_trim_wasm_always

# Conflicts:
#	src/mono/wasm/Wasm.Build.Tests/WasmTemplateTestBase.cs

maraf commented:
@fanyang-mono @radical WBT now pass 🎉 Let me know if the SDK changes need anything

fanyang-mono committed:
Add a blazor template test

fanyang-mono commented:
@radical Please review my latest bug fix (https://github.com/dotnet/runtime/pull/90436/commits/1c9e04d19655b45a862ff7b6aa27d90b44470ab2) and the two new tests.

fanyang-mono commented:
/azp run runtime-wasm

radical commented on src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs:
```diff
@@ -491,5 +491,68 @@ public void BuildAndRunForDifferentOutputPaths(string config, bool appendRID, bo
+            Assert.True(Directory.Exists(origAssemblyDir), $"Could not find the original AOT input assemblies dir: {origAssemblyDir}");
+            Assert.True(Directory.Exists(strippedAssemblyDir), $"Could not find the stripped assemblies dir: {strippedAssemblyDir}");
+
+            string assemblyToExam = "System.Private.CoreLib.dll";
```
```suggestion
            string assemblyToExamine = "System.Private.CoreLib.dll";
```

radical commented on src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs:
```diff
@@ -491,5 +491,68 @@ public void BuildAndRunForDifferentOutputPaths(string config, bool appendRID, bo
+            Assert.True(File.Exists(strippedAssembly));
+            Assert.True(File.Exists(bundledAssembly));
+
+            string compressedOriginalAssembly = Compress(originalAssembly);
```
Why do we need to compare the compressed sizes?

fanyang-mono replied:
Because the .dll files are the same size. The trim was done by replacing method body by zeros.

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -305,12 +305,63 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
     {
+        string monoAotPropertyValuesFilePath = Path.Combine(IntermediateOutputPath, "monoAotPropertyValues.txt");
+
+        if (File.Exists(monoAotPropertyValuesFilePath))
```
- move this to a separate method, which just returns the dictionary
- and here after the call, you can check for the specific property
- This would allow use of this mechanism for other properties in future

fanyang-mono replied:
done

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -305,12 +305,63 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
+                File.Delete(monoAotPropertyValuesFilePath);
+        }
+
+        if (!File.Exists(monoAotPropertyValuesFilePath) || ((_collectTrimmingEligibleMethodsValueFromPreviousBuild != CollectTrimmingEligibleMethods)))
```
- separate method that is not specific to `CollectTrimmingEligibleMethods`

fanyang-mono replied:
done

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -305,12 +305,63 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
+                Log.LogMessage(MessageImportance.High, $"Deleting {file.Name} to force a new AOT compilation, because the value of CollectTrimmingEligibleMethods has changed");
+                file.Delete();
+            }
+            foreach (FileInfo file in di.GetFiles("*.o"))
```
This will get *any* `*.o` files in `obj/`, and not just the ones related to aot.

fanyang-mono replied:
Originally, I didn't intend to remove the `.o` files, as they are not the output of the aot task. However, I ran into a problem where the task (`EmccCompile`) to generate the `.o` files would generated a new `.o` file but would not replace with the old one, as the content is the same. Example,
```
Did not overwrite /Users/yangfan/Documents/work/dotnet_1/runtime/artifacts/obj/mono/Wasm.Console.V8.Sample/wasm/Release/browser-wasm/wasm/for-publish/Wasm.Console.V8.Sample.dll.o as the contents are unchanged
```
That caused a rebuild issue afterwards, because the `.o` files were older than the `.bc` files.
In order to work around this issue, I had to remove the `.o` files here as well.

Maybe, the correct thing to do here is to always overwrite newly generated `.o` files in `EmccCompile`? I am not sure what side effect that might bring.

radical replied:
This task should not remove files that were not generated by itself.

fanyang-mono replied:
You are right. Will update `EmccCompile` to always overwrite. Will see if it breaks any existing behavior.

fanyang-mono replied:
The change of `EmccCompile`  triggered `Wasm.Build.NativeRebuild.Tests.FlagsChangeRebuildTests.ExtraEmccFlagsSetButNoRealChange` to fail. We have to use cases which require conflicting changes. What shall we do now?

radical replied:
> Originally, I didn't intend to remove the `.o` files, as they are not the output of the aot task. However, I ran into a problem where the task (`EmccCompile`) to generate the `.o` files would generated a new `.o` file but would not replace with the old one, as the content is the same. Example,
> 
> ```
> Did not overwrite /Users/yangfan/Documents/work/dotnet_1/runtime/artifacts/obj/mono/Wasm.Console.V8.Sample/wasm/Release/browser-wasm/wasm/for-publish/Wasm.Console.V8.Sample.dll.o as the contents are unchanged
> ```

that is the correct behavior.

> That caused a rebuild issue afterwards, because the `.o` files were older than the `.bc` files. In order to work around this issue, I had to remove the `.o` files here as well.

Which test was failing because of this?

> Maybe, the correct thing to do here is to always overwrite newly generated `.o` files in `EmccCompile`? I am not sure what side effect that might bring.

That would always trigger rebuilds then. I had solved a similar issue before, but need more details.

fanyang-mono replied:
Discussed offline.

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -305,12 +305,63 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
+        }
+
+        if (_collectTrimmingEligibleMethodsValueFromPreviousBuild != CollectTrimmingEligibleMethods)
+        {
```
I need to think about what we are doing here.

radical reviewed (dismissed):
On-going review

fanyang-mono committed:
Address review feedback

fanyang-mono committed:
Update src/mono/wasm/Wasm.Build.Tests/Templates/WasmTemplateTests.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

fanyang-mono committed:
Update parameter name

fanyang-mono committed:
Update usage of parameter

radical commented:
I would suggest splitting this PR into:

1. the `mono/mono`, and `src/libraries` changes
2. `RuntimeConfigJsonPath` change, and the bit in Browser.targets
3. Rest of the changes

fanyang-mono committed:
Revert EmccCompile change

fanyang-mono commented:
/azp run runtime-wasm

radical committed:
Merge remote-tracking branch 'origin/main' into enable_il_trim_wasm_always

radical committed:
MonoAOTCompiler: revert rebuilding when WasmStripILAfterAOT value changes

radical committed:
Compress -> GZipCompress, and cleanup

radical committed:
rework the tests

- to support webcil case
- run blazor app
- add cases for the default setting, and the opposite
- cleanup

radical committed:
ILStrip.cs: Emit a message about stripping to make it obvious to the user

radical commented:
As discussed offline, I have pushed the following changes:
- reverted the rebuild-if-the-property-changes
- the tests were explicitly disabling webcil, fixed that to work with webcil, though the tests compare the file size which is not useful for webcil case. In a future PR we should check for stripping in some other way
- Added another case for testing the non-default value also

radical committed:
WasmApp.targets: update comment

radical commented:
/azp run runtime-wasm

radical reviewed (approved):
Thank you for your patience! LGTM 👍

fanyang-mono committed:
Change default value to false and update test

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono committed:
Merge remote-tracking branch 'origin/main' into enable_il_trim_wasm_always

fanyang-mono commented:
/azp run runtime-wasm

fanyang-mono commented:
According too `Build Analysis`, all test failures are known.
